lib/acme/client.rb

```diff
@@ -41,93 +39,221 @@ def initialize(jwk: nil, private_key: nil, endpoint: DEFAULT_ENDPOINT, directory
       Acme::Client::JWK.from_private_key(private_key)
     end
 
-    @endpoint, @directory_uri, @connection_options = endpoint, directory_uri, connection_options
+    @kid, @directory_url, @connection_options = kid, URI(directory), connection_options
     @nonces ||= []
-    load_directory!
   end
 
-  attr_reader :jwk, :nonces, :endpoint, :directory_uri, :operation_endpoints
+  attr_reader :jwk, :nonces, :directory
 
-  def register(contact:)
+  def new_account(contact:, terms_of_service_agreed: nil)
     payload = {
-      resource: 'new-reg', contact: Array(contact)
+      contact: Array(contact)
     }
 
-    response = connection.post(@operation_endpoints.fetch('new-reg'), payload)
-    ::Acme::Client::Resources::Registration.new(self, response)
+    if terms_of_service_agreed
+      payload[:termsOfServiceAgreed] = terms_of_service_agreed
+    end
+
+    response = post(endpoint_for('newAccount'), payload: payload, mode: :jws)
+    @kid = response.headers.fetch(:location)
+
+    if response.body.nil? || response.body.empty?
```

Comment:
https://tools.ietf.org/html/draft-ietf-acme-acme-11#section-7.3.1

If an account is created, the server responds with a `201`, if not a `200`.

Reply:
I know its a bit weird but [pebble](https://github.com/letsencrypt/pebble) and [boulder](https://github.com/letsencrypt/boulder) are no implemented exactly the same. If I remember correctly pebble return empty body on `201` and expect you to follow the location.

I found that the presence of the location was consistent whenever the body was empty, so that's why I did this. Ideally, the body should always be present.

Reply:
^ @cpu I'm getting this right?

Reply:
@unixcharles This is in the context of a normal newAccount request correct, not one with `onlyReturnExisting: true`? I don't think this is right if so.

Pebble returns HTTP status `201` **and** the account object for successful `newAccount` requests that create a new account: https://github.com/letsencrypt/pebble/blob/cae1c3522ee7a4871550f4ff9378677e06a4cdf7/wfe/wfe.go#L739

Boulder does the same thing: HTTP status 201 and the account object in the response body: https://github.com/letsencrypt/boulder/blob/3ecf841a3a48b7c504f26e5b040fbfb36c1664b7/wfe2/wfe.go#L531

Reply:
I think that it happened when attempting to call `newAccount` on an account that was already registered without `onlyReturnExisting: true`.

I'll try to reproduce, maybe I was just doing something wrong with my test setup.

Reply:
> I think that it happened when attempting to call newAccount on an account that was already registered without onlyReturnExisting: true.

That would make sense :+1: - There should not be an HTTP Status Created response in this case because no account was created. That also matches what Section 7.3.1 says when you're not using `onlyReturnExisting`:

     If the server already has an account registered with the provided
     account key, then it MUST return a response with a 200 (OK) status
     code and provide the URL of that account in the Location header
     field.  This allows a client that has an account key but not the
     corresponding account URL to recover the account URL.

In that specific case Boulder returns no body and HTTP status OK: https://github.com/letsencrypt/pebble/blob/cae1c3522ee7a4871550f4ff9378677e06a4cdf7/wfe/wfe.go#L686-L693 Pebble does the same: https://github.com/letsencrypt/pebble/blob/cae1c3522ee7a4871550f4ff9378677e06a4cdf7/wfe/wfe.go#L686-L693

Reply:
Okay, so it would make sense to follow the location headers if not body is returned.